### PR TITLE
Dynamic Options - Implement new helper functions & update widget-load behaviour

### LINKED ISSUE #1122: Re-architect widget-load and data transfer workflow
### Description

Capturing this in a new issue after lengthy discussions and issues captured here:

- https://discourse.nodered.org/t/how-to-sync-the-full-config-between-server-and-client-side-state/89252/80
- https://github.com/FlowFuse/node-red-dashboard/issues/1117

#### Widget Load Loses State

The problem can be seen in the Events Architecture [here](https://dashboard.flowfuse.com/contributing/guides/events.html#loading-event-flow):

Screenshot 2024-07-24 at 14 22 57

Note that in the `widget-load` event, there is no reference to the `state` store, so when a user navigates away from a page, updates to that store can occur, but the `widget-load` events is not aware of them.

Loosely thinking, we need to have something like:

Screenshot 2024-07-24 at 14 23 39

#### Data Store Changes

There is also merit here in swapping how the `datastore` is used - rather than being a full `msg` store - we could treat it like `state` in that it's a singular, cumulative reference to the `data` for a single widget. For example, if `msg.payload` and `msg.otherProperty` are sent in _different_ messages, they would still be stored in the same place, and would be replayed accordingly.

### PR DESCRIPTION
**Pull request by joepavitt**
## Description

Okay, this is a fairly big shift, so will do my best to articulate the changes, and the _why_.

### `widget-load` changes

- We now include a `state` value in the `widget-load` event. This is because any changes, and dynamic properties, set on a node when not in view on the Dashboard would be updated server-side, but not client-side. So, when a user navigates (via the side navigation) to the relevant page, the state loaded is still the original configuration provided by Node-RED's latest deploy.
- In the client-side handler for this event, we now automatically put everything received in `state` straight into our VueX store under the `widget.state` field. This means that we don't need to define this on _every_ widget, which we would need to. Note though, that we still need to have the server-side definitions over what can go into `state`, and the client-side widgets need to reference this still via `computed` variables (explained next)
- This event will now fire for _every_ widget, as long as it has `msg` _or_ `state` saved in the server-side stores. Previously, it would only fire if something was present in `datastore` (i.e. last received `msg`)

### Move to a centralised store model

- I've removed the use of the `dynamic` design pattern that we had implemented in order to track the dynamic overrides. 
- Instead, we're referencing the `widget` state stored in vuex directly using `computed` variables. This ensures we have a single source of truth, and not many copies of the truth distributed out to each widget/node.

### New global helper functions

- In order to make the interface to the vuex store easier, I've created two new global helper functions for dealing with dynamic (and static) properties:
    - `updateDynamicProperty(property, value)`: Updates a property with the provided value, and checks against undefined types
    - `setDynamicProperties(config)`: Will set the provided properties in the vuex store for a given widget this is run on. This will automatically update the widget's state, and any references using this property.
    - `getProperty(property)`: Automatically gets the correct value for the requested property. Will first look in the dynamic properties, and if not found, will default to the static configuration (i.e. what's defined in Node-RED) that arrived via the `ui-config` event.

### `widgetState` store mutation changes

- The arguments for the `widgetState` mutation on the `ui` vuex store has changed to permit separation of the proeprties being set and the `id` of the widget.

## Progress

We need to sweep over all of the existing widgets that were already using the design pattern to ensure consistency:

- [x] UI Form
- [x] UI Button
- [x] UI Button Group
- [x] UI Dropdown
- [x] UI Slider

We can then continue https://github.com/FlowFuse/node-red-dashboard/issues/833 with the new pattern introduced here

## Related Issue(s)

Closes https://github.com/FlowFuse/node-red-dashboard/issues/793
Closes https://github.com/FlowFuse/node-red-dashboard/issues/1022
Closes https://github.com/FlowFuse/node-red-dashboard/issues/1117
Closes #1122

Need to double check once closed on whether it helps https://github.com/FlowFuse/node-red-dashboard/issues/871

### CODE REVIEW

**Comment by joepavitt** [comment]
Only ui-dropdown to go, which is a little trickier as has some backward compatibility stuff left over with `msg.options` that we still need to support - will attack that tomorrow

File: ui/src/widgets/ui-button/UIButton.vue [hunk 1]
```diff
@@ -93,26 +81,23 @@ export default {
...
+                this.setDynamicProperties({ buttonColor: updates.buttonColor })
             }
             if (typeof updates.textColor !== 'undefined') {
-                this.dynamic.textColor = updates.textColor
+                this.setDynamicProperties({ textColor: updates.textColor })
             }
             if (typeof updates.iconColor !== 'undefined') {
-                this.dynamic.iconColor = updates.iconColor
+                this.setDynamicProperties({ iconColor: updates.iconColor })
             }
```
Comment (Steve-Mcl) [inline ui/src/widgets/ui-button/UIButton.vue:100]: Joe, I think we can both cement and simplify this behaviour with an additional helper. 

### Additional helper, with guards and the `undefined` pattern built in for 100% consistency

```
                updateDynamicProperty(prop, config) {
                    if (!prop || !config || typeof config !== 'object') {
                       return // maybe better to throw error here as this is likely program error
                    }
                    if (typeof config[prop] === 'undefined') {
                       return // no change
                    }
                    this.$store.commit('ui/widgetState', {
                        widgetId: this.id,
                        config
                    })
                },
```

### Now, all calls, by all nodes (contrib too) become consistent
```
          this.updateDynamicProperty('label', updates)
          this.updateDynamicProperty('icon', updates)
          this.updateDynamicProperty('iconPosition', updates)
          this.updateDynamicProperty('buttonColor', updates)
          this.updateDynamicProperty('textColor', updates)
          this.updateDynamicProperty('iconColor', updates)
```


---

If necessary (i.e there is a reason to permit `undefined` as a value) we can add an options object to the `updateDynamicProperty` e.g.  `this.updateDynamicProperty('label', updates, { allowUndefined: true })`

I realise this is a little extra work but as this is a large change already, we have the opportunity to cement good practice at the core.

Happy to discuss over a call.

**Comment by colinl** [comment]
How are validations being done on ui_update operations, to make sure invalid data are not allowed?

**Comment by joepavitt** [comment]
> How are validations being done on ui_update operations, to make sure invalid data are not allowed?

Needs to be done on a node-by-node basis. Right now, not much validation.

Also PR opened for UI Example updates: https://github.com/FlowFuse/node-red-dashboard-2-ui-example/pull/17 will get it reviewed once these core PRs are merged

**Comment by joepavitt** [comment]
> One question regarding the use of the dynamic properties, is it sensible that the user may want to return a specific property to its default value, (for example button color, that was defined in the config window as yellow, or the label text) by sending a null or an undefined?

If the value is set to `null`, then the property is reset. Our `getProperty` logic checks:

```js
return this.state && property in this.state && state !== null ? state : config
```

Note, if `state === null`, then the original `config` will be returned

**Comment by arturv2000** [comment]
I was testing that scenario, and it doesn't work.

For example:
- I have a button that I defined in the config as the color is `yellow` and perform a full-deploy.
- I send a message setting the color to `red` using `ui_update.buttonColor`
- In this case, if I send `ui_update.buttonColor = null` the button color will return to `yellow`
- If I refresh the page, the button color return to the theme default, if send again the `ui_update.buttonColor = null` it has no effect. If before I refresh the page i set the color to `green` after the refresh sending a `null` will return the color to `green`.

Not a big issue from my part, but may be something to consider if it is necessary to have a way to revert to initial configuration.

**Comment by joepavitt** [comment]
Will follow up with outstanding nodes in https://github.com/FlowFuse/node-red-dashboard/issues/833 once this is merged

**Comment by Steve-Mcl** [comment]
### ui-dropdown

SQLiteStudio_mYtMWIPAd0


1. Setting label dynamically works but reverting by sending null doesnt appear to work as exepcted
2. Dynamically setting multiple `true`/`false` causes odd entries in the current selection



chrome_Yirme9hgqh

3. After applying dynamic label a refresh forgets selection

**Comment by joepavitt** [comment]
UI Dropdown:

1. We don't support that currently, will be a follow up exercise.

2. If you're referring to the vertical alignment problem, that's fixed in a PR that Gayan opened yesterday, it was an underlying Vuetify bug that's fixed in newest version.

3. I'd expect selection to be forgotten anyway on refresh? Unless the last message received contains the `payload` it's a shortfall of the "replay last message" approach we've copied from Dashboard 1.0

**Comment by Steve-Mcl** [comment]
> 2\. If you're referring to the vertical alignment problem, that's fixed in a PR that Gayan opened yesterday, it was an underlying Vuetify bug that's fixed in newest version

No, i am referring to what happens in the GIF ("two" is selected, I inject `msg.ui_update.multiple = true`, "two two" is now shown!)
image
I posted the demo flow too.


> 1. We don't support that currently, will be a follow up exercise.

Don't support dynamic label for ui-dropdown? 
As in not documented right? 
_(because is does work and there is code for it in this PR - so I tested it)_




> 3\. I'd expect selection to be forgotten anyway on refresh? Unless the last message received contains the `payload` it's a shortfall of the "replay last message" approach we've copied from Dashboard 1.0

Is this something we can fix/improve?

**Comment by joepavitt** [comment]
> Don't support dynamic label for ui-dropdown?
As in not documented right?

The use of `null` to clear it, which is what I thought is what you're saying was not working?

**Comment by joepavitt** [comment]
Now reliving what you mean re: multiple. That's likely an issue in current `main`, I'm AFK but if you're able to check that'd be great.

Don't think that's a problem been introduced here

**Comment by joepavitt** [comment]
> Is the problem here that a message containing only ui_update settings is stored as the last message? I think that such messages should not be stored, as the ui_update data are stored in the state store.

This could work, but not in scope of this issue

**Comment by joepavitt** [comment]
> I thought the point of making helper functions was to cement this common behaviour at the core

It's not listed as a supported or new feature, so no, this has lot been added, but can be in a follow up PR/issue

**Comment by Steve-Mcl** [comment]
> > I thought the point of making helper functions was to cement this common behaviour at the core
> 
> It's not listed as a supported or new feature, so no, this has lot been added, but can be in a follow up PR/issue

Are any of these changes breaking or will users experience differing behaviour when this is merged Joe?  If yes, then I would personally recommend we get the behaviour nailed down in this PR before we release it into the wild.

**Comment by Steve-Mcl** [comment]
#### ui-number-group

chrome_qTk75XCTdo


1. Sending payload with matching value does not slect the matching button. After a refresh it does become selected.
2. Sending an updated label correctly sets the label.   Sending `null` correctly reverts the label to static value. However, if the DB is refreshed, sending `null` no longer reverts label to static value.
3. Sending a new set of options with same/matching `value`s does not keep the current selection.

demo flow

```
[{"id":"8c7ef62c4af9d424","type":"inject","z":"f80624abfd69b02e","name":"ui_update.label Numbers","props":[{"p":"ui_update.label","v":"Numbers","vt":"str"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1250,"y":80,"wires":[["75551956e485bac6"]]},{"id":"285f16e64f307827","type":"inject","z":"f80624abfd69b02e","name":"ui_update.label null","props":[{"p":"ui_update.label","v":"null","vt":"json"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1270,"y":120,"wires":[["75551956e485bac6"]]},{"id":"48af77c3c271917d","type":"inject","z":"f80624abfd69b02e","name":"ui_update.options numbers","props":[{"p":"ui_update.options","v":"[{\"value\":1,\"label\":\"One\",\"icon\":\"numeric-positive-1\"},{\"value\":true,\"label\":\"Two\",\"icon\":\"numeric-2-box\"},{\"value\":\"tree\",\"label\":\"T_ree\",\"icon\":\"tree\"}]","vt":"json"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1250,"y":200,"wires":[["75551956e485bac6"]]},{"id":"7a50a8231976b836","type":"inject","z":"f80624abfd69b02e","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"1","payloadType":"num","x":1310,"y":240,"wires":[["75551956e485bac6"]]},{"id":"17dc21e121c2184c","type":"inject","z":"f80624abfd69b02e","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"tree","payloadType":"str","x":1310,"y":320,"wires":[["75551956e485bac6"]]},{"id":"cb13084236dfcfc8","type":"inject","z":"f80624abfd69b02e","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"true","payloadType":"bool","x":1310,"y":280,"wires":[["75551956e485bac6"]]},{"id":"80798abd6bac3521","type":"debug","z":"f80624abfd69b02e","name":"debug 187","active":true,"tosidebar":true,"console":false,"tostatus":true,"complete":"payload","targetType":"msg","statusVal":"payload","statusType":"auto","x":1540,"y":200,"wires":[]},{"id":"8021da3ecf8e21f9","type":"inject","z":"f80624abfd69b02e","name":"ui_update.options null","props":[{"p":"ui_update.options","v":"null","vt":"json"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1260,"y":360,"wires":[["75551956e485bac6"]]},{"id":"2c1654b9f7489958","type":"ui-button-group","z":"f80624abfd69b02e","name":"","group":"5c9407416fba1d1b","order":1,"width":6,"height":1,"label":"stars","rounded":false,"useThemeColors":true,"passthru":false,"options":[{"label":"Cap'n A","icon":"star-circle-outline","value":"1","valueType":"num","color":"#009933"},{"label":"Crescent","icon":"star-crescent","value":"true","valueType":"bool","color":"#999999"},{"label":"David","icon":"star-david","value":"tree","valueType":"str","color":"#ff6666"}],"topic":"topic","topicType":"msg","x":1530,"y":140,"wires":[["80798abd6bac3521"]]},{"id":"75551956e485bac6","type":"junction","z":"f80624abfd69b02e","x":1440,"y":140,"wires":[["2c1654b9f7489958"]]},{"id":"5c9407416fba1d1b","type":"ui-group","name":"PR-1123","page":"868aaffa6da2e1ce","width":"6","height":"1","order":1,"showTitle":true,"className":"","visible":"true","disabled":"false"},{"id":"868aaffa6da2e1ce","type":"ui-page","name":"PR-1123","ui":"22ea43815413e748","path":"/page13","icon":"home","layout":"grid","theme":"c2ff5ba1f92a0f0e","order":1,"className":"","visible":"true","disabled":"false"},{"id":"22ea43815413e748","type":"ui-base","name":"base","path":"/dashboard","includeClientData":true,"acceptsClientConfig":["ui-notification","ui-control"],"showPathInSidebar":false,"navigationStyle":"default","titleBarStyle":"default"},{"id":"c2ff5ba1f92a0f0e","type":"ui-theme","name":"Default","colors":{"surface":"#ffffff","primary":"#0094ce","bgPage":"#eeeeee","groupBg":"#ffffff","groupOutline":"#cccccc"},"sizes":{"pagePadding":"12px","groupGap":"12px","groupBorderRadius":"4px","widgetGap":"12px"}}]
```

**Comment by joepavitt** [comment]
> Are any of these changes breaking or will users experience differing behaviour when this is merged Joe?

There should be 0 difference to UX as a result of this PR

**Comment by Steve-Mcl** [comment]
> > Are any of these changes breaking or will users experience differing behaviour when this is merged Joe?
> 
> There should be 0 difference to UX as a result of this PR

Ok, well in which case my semi - thorough tests should probably be halter until I can establish a baseline to see if this Pr is causing the issues I see or if they are existing!


Before I halt, I will post findings for UI-Button (rather than let my findings go to waste)

#### ui-button

chrome_7lXqVChiYj

1. Setting dynamic label works. sending `null` for label reverts it. However, after a refresh, it can not be reverted anymore.
2. Setting dynamic icon + iconPosition works. sending `null` for icon + iconPosition reverts them. However, after a refresh, they can no longer be reverted.

chrome_o65l3ZoRzM

3.  Setting dynamic label works. sending `null` for label reverts it. However if we refresh with the label reverted via a `null`, the button has no label after a refresh
4. Setting dynamic icon + iconPosition works. sending `null` for icon + iconPosition reverts them. However, if we refresh with the icon + iconPosition revered via a `null`, the button is lost. An error in console reveals the issue is in `makeMdiIcon`
```
TypeError: Cannot read properties of null (reading 'replace')
    at Proxy.makeMdiIcon (index-BPCbQ1sF.js:47:36548)
    at Proxy.prependIcon (index-BPCbQ1sF.js:47:35741)
    at ReactiveEffect.fn (index-BPCbQ1sF.js:10:13076)
    at ReactiveEffect.run (index-BPCbQ1sF.js:10:1918)
    at get value (index-BPCbQ1sF.js:10:13345)
    at Object.get [as prependIcon] (index-BPCbQ1sF.js:17:34086)
    at Proxy._sfc_render$h (index-BPCbQ1sF.js:47:37151)
    at renderComponentRoot (index-BPCbQ1sF.js:17:7654)
    at ReactiveEffect.Ce [as fn] (index-BPCbQ1sF.js:17:54912)
    at ReactiveEffect.run (index-BPCbQ1sF.js:10:1918)
```
image


demo flow
```
[{"id":"a19fe7752981edcb","type":"inject","z":"f80624abfd69b02e","name":"icon & position left","props":[{"p":"ui_update.icon","v":"menu-left","vt":"str"},{"p":"ui_update.iconPosition","v":"left","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1170,"y":80,"wires":[["225e4e1efcbc0ddc"]]},{"id":"019c68b113848730","type":"inject","z":"f80624abfd69b02e","name":"icon & position null","props":[{"p":"ui_update.icon","v":"null","vt":"json"},{"p":"ui_update.iconPosition","v":"null","vt":"json"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1170,"y":120,"wires":[["225e4e1efcbc0ddc"]]},{"id":"303170327edd25aa","type":"inject","z":"f80624abfd69b02e","name":"ui_update.label BTN TWO","props":[{"p":"ui_update.label","v":"BTN TWO","vt":"str"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1150,"y":240,"wires":[["1e59de55be3906d2"]]},{"id":"835d0f44edce12c9","type":"inject","z":"f80624abfd69b02e","name":"ui_update.label null","props":[{"p":"ui_update.label","v":"null","vt":"json"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1170,"y":280,"wires":[["1e59de55be3906d2"]]},{"id":"97fc0235988ce987","type":"debug","z":"f80624abfd69b02e","name":"debug 188","active":true,"tosidebar":true,"console":false,"tostatus":true,"complete":"payload","targetType":"msg","statusVal":"payload","statusType":"auto","x":1440,"y":180,"wires":[]},{"id":"225e4e1efcbc0ddc","type":"ui-button","z":"f80624abfd69b02e","group":"5c9407416fba1d1b","name":"Button 1","label":"Button 1","order":1,"width":"3","height":"1","emulateClick":false,"tooltip":"","color":"","bgcolor":"","className":"","icon":"menu-right","iconPosition":"right","payload":"1","payloadType":"num","topic":"topic","topicType":"msg","buttonColor":"","textColor":"","iconColor":"","x":1380,"y":100,"wires":[["97fc0235988ce987"]]},{"id":"1e59de55be3906d2","type":"ui-button","z":"f80624abfd69b02e","group":"5c9407416fba1d1b","name":"Button 2","label":"Button 2","order":2,"width":"3","height":"1","emulateClick":false,"tooltip":"","color":"","bgcolor":"","className":"","icon":"","iconPosition":"right","payload":"two","payloadType":"str","topic":"topic","topicType":"msg","buttonColor":"","textColor":"","iconColor":"","x":1380,"y":260,"wires":[["97fc0235988ce987"]]},{"id":"5c9407416fba1d1b","type":"ui-group","name":"PR-1123","page":"868aaffa6da2e1ce","width":"6","height":"1","order":1,"showTitle":true,"className":"","visible":"true","disabled":"false"},{"id":"868aaffa6da2e1ce","type":"ui-page","name":"PR-1123","ui":"22ea43815413e748","path":"/page13","icon":"home","layout":"grid","theme":"c2ff5ba1f92a0f0e","order":1,"className":"","visible":"true","disabled":"false"},{"id":"22ea43815413e748","type":"ui-base","name":"base","path":"/dashboard","includeClientData":true,"acceptsClientConfig":["ui-notification","ui-control"],"showPathInSidebar":false,"navigationStyle":"default","titleBarStyle":"default"},{"id":"c2ff5ba1f92a0f0e","type":"ui-theme","name":"Default","colors":{"surface":"#ffffff","primary":"#0094ce","bgPage":"#eeeeee","groupBg":"#ffffff","groupOutline":"#cccccc"},"sizes":{"pagePadding":"12px","groupGap":"12px","groupBorderRadius":"4px","widgetGap":"12px"}}]
```


<br>

---

What I will state is, there is a common theme across many of the issues related to override/null/refresh issues

**Comment by arturv2000** [comment]
Most of these issues exists in Main, regarding the `null`/reverting to original configuration issue was already talked about in this PR, and was decided to address that latter on.
https://github.com/FlowFuse/node-red-dashboard/pull/1123#issuecomment-2258110125

**Comment by arturv2000** [comment]
Regarding the `ui-number-group` it is a `bug` from this PR

before was this:
```js
const updates = msg.ui_update
if (typeof updates?.label !== 'undefined') {
   this.dynamic.label = updates.label
}
if (typeof updates?.options !== 'undefined') {
   this.dynamic.options = updates.options
}
```

And now is:
```js
const updates = msg.ui_update
this.updateDynamicProperty('label', updates.label)
this.updateDynamicProperty('options', updates.options)
```

But when we inject a message via Node-Red, the `onDynamicProperties` is called, but the object `ui_update` may not exits, or the property `updates.label` may not exit...


Replacing with:
```js
        onDynamicProperty (msg) {
            const updates = msg.ui_update
            if (!updates) {
                return
            }
            this.updateDynamicProperty('label', updates?.label)
            this.updateDynamicProperty('options', updates?.options)
        },
```
Solves the main issue (Not the null not returning to original configuration)

The same behaviour may be happening in other widgets

**Comment by joepavitt** [comment]
So, I'm fairly sure this is all good to go now. @Steve-Mcl the `null` concerns will be treated as a follow up feature request, and are not a blocker for this PR

**Review by Steve-Mcl — APPROVED** [review]
So the newly introduced issues are indeed resolved. 

Code review all good.

NOTE: This is approved in the knowledge that there are a few issues related to mostly to setting then reseting dynamic props but they are existing and are to be addressed separately.  

In case they are not all spelled out in the issue, the ones I seen are:

* dropdown
    * Applying dynamic label then resetting it with a `null` followed by a refresh causes the label to disappear
    * Applying dynamic options then resetting them with a `null` followed by a refresh causes the options to disappear

* form
    * Applying dynamic label then resetting it with a `null` followed by a refresh causes the label to disappear
    * Applying dynamic options then resetting them with a `null` followed by a refresh causes the options to disappear
    * After Applying dynamic options, the first 2 fields appear to be bound (typing in one field will update the other)

* button
    * Applying dynamic icon then resetting it with a `null` followed by a refresh causes the button to disappear
    * Applying dynamic label then resetting it with a `null` followed by a refresh causes the buttons label to disappear

* button group
    * Applying dynamic label then resetting it with a `null` followed by a refresh causes the label to disappear
    * Applying dynamic options then resetting them with a `null` followed by a refresh causes the options to disappear


 NOTE: This ↑ was not exhaustive - only a selection of the dynamic ops and widgets were tested for the purposes of this review.

**Comment by joepavitt** [comment]
Thanks Steve

**Comment by omrid01** [comment]
So just to make sure, there is no backwards compatibility impact:
- server node will continue to send 'on-widget-load' to new clients, with their respective data from the datastore?
- Will on-widget-load now be sent also when the datastore has no data for this node?

**Comment by joepavitt** [comment]
> server node will continue to send 'on-widget-load' to new clients, with their respective data from the datastore?

correct

> Will on-widget-load now be sent also when the datastore has no data for this node?

Yes, correct

**Comment by omrid01** [comment]
Excellent, thank you.
Is there a way to know the dash-2.0 version (programmatically from within the server node?)

**Comment by joepavitt** [comment]
Not that I can think of - any reason you'd need to know for this though, in your `load` handler, just account for `if (!msg)` being a ting, and it should be backward compatabile.

**Comment by omrid01** [comment]
I'm trying to think ahead. Dash-2.0 is continuously evolving with new functionality, and it would be good to enable a node to know the dash-2.0 version it is running on (and determine which framework functionality is available). Equivalent to checking the NodeJS version to determine if a new JavaScript feature is available.

Specifically in my case, it's chicken & egg. I can easily check if a message is null or undefined once I receive the load notification, but will never get this notification in the first place if datastore was empty... :-)  so knowing the version will enable me to retain the current workaround (forcing a dummy datastore entry) for users who upgrade the node but are still on an older dashboard version.

**Pull request merged**